DB-27 Add Leaves

Sirus97 opened the pull request.

Sirus97 committed:
Add Leaves

Sirus97 committed:
Better variables

MarkoDedovic commented on src/main/handlers/leaves.py:
```diff
@@ -0,0 +1,70 @@
+
+    async def execute(self, message):
+        if not message.guild.id in self.server_vars:
+            self.server_vars[message.guild.id] = {}
```
why do these values need to be stored in a dict? it seems simpler to just put them in their own variables

Sirus97 replied:
It's set up to handle several Discord servers so the different values need to be stored somewhere else.

MarkoDedovic commented on src/main/handlers/leaves.py:
```diff
@@ -0,0 +1,70 @@
+            self.server_vars[message.guild.id][TIMES_REPEAT] = 5
+            self.server_vars[message.guild.id][LAST_LEAVES] = datetime.min
+
+        if self.server_vars[message.guild.id][ONGOING]:
```
instead of checking if a leaves is happening on every message, why not start a loop on the initial trigger? `for i in range(TIMES_REPEAT):`

Sirus97 replied:
A loop would most likely create a column of leaves instead of the intended function which is a new leaf after some messages have been sent to appear as if it's falling through the chat.

MarkoDedovic commented on src/main/handlers/leaves.py:
```diff
@@ -0,0 +1,70 @@
+                print('Another leaves is currently still going!')
+            else:
+                print(f'{datetime.now().isoformat()} Throwing leaves!')
+                self.server_vars[message.guild.id][TIMES_REPEAT] = random.choices((random.choice(range(1, 10)), 30), weights=[95, 5])[0]
```
what's the point of a random choice on a random choice?

Sirus97 replied:
Could improve the readability here if you think it's justified. The internal `random.choice` chooses a number from 1 through 10 with equal weight and the external `random.choices` chooses between that number and a 5% chance of repeating a ridiculous number of times, 30 in this case.

Sirus97 committed:
Trim trailing whitespace

Sirus97 committed:
Fix bugs
- Only count messages sent in the same channel
- Only send messages in the same channel
- Only throw in channels with permission to send messages

Sirus97 committed:
Give every channel a separate Leaves instance

Sirus97 committed:
Reset leaves sent counter after done throwing

Sirus97 committed:
Add config option for Leaves cooldown time

Sirus97 committed:
Delay message counter reset

Sirus97 commented:
Don't merge this yet.
